### ISSUE TITLE
[Question] How to render custom tags?

### ISSUE THREAD

**Issue by thomthom** [lsegal/yard#1227]
I'm aware of the basic custom tag config switches:

```
    # * +--tag+ TAG:TITLE
    # * +--name-tag+ TAG:TITLE
    # * +--type-tag+ TAG:TITLE
    # * +--type-name-tag+ TAG:TITLE
    # * +--title-tag+ TAG:TITLE
```

1. I'm however unclear what the difference is. I didn't find any examples in the documentation. Did I miss it?

2. If I want to customize the rendering of a given tag, how to I go about doing that?

I see some logic in https://github.com/lsegal/yard/blob/master/templates/default/tags/setup.rb

`init` appear to call `create_tag_methods` on visible tags with some exceptions.

`create_tag_methods` appear to create method names with the same name as the tag (?) if a method of that name doesn't already exist. Does that mean I can create a custom `templates/default/tags/setup.rb` with the names of my custom tags and that will then be called when the tag is rendered?

Or is there a different way to deal with this?

**Comment by thomthom** [comment]
I tried to replace the config defined tag: `--tag bug:"Known Bugs"` with this:

```
YARD::Tags::Library.define_tag('Known Bugs', :bug)
```

But that doesn't seem to be the same. The tag doesn't appear in the rendered output.

**Comment by lsegal** [comment]
> 1. I'm however unclear what the difference is. I didn't find any examples in the documentation. Did I miss it?

This is a noted documentation blindspot. PR would be accepted to add this documentation somewhere, but in short:

* -tag - bare freeform text tag
* -name-tag - freeform text tag attached to a key name-- think of `@see` which has a "name" and "text" component (name being the URL, text being description)
* -type-tag - freeform text tag with a type specifier parsed before the tag -- i.e., `@return [TYPES] text`
* -type-name-tag - freeform text tag with a type specifier and key name, i.e. `@param name [TYPES] text`
* -title-tag - freeform text tag attached to a "title" -- this is similar to a name-tag except the "key" portion allows for spaces and the title / text are separated by a newline. This is used for `@example` where you want a longer "name" specifier than a single word.

> 2\. If I want to customize the rendering of a given tag, how to I go about doing that?

You're on the right track below here:

> `create_tag_methods` appear to create method names with the same name as the tag (?) if a method of that name doesn't already exist. Does that mean I can create a custom `templates/default/tags/setup.rb` with the names of my custom tags and that will then be called when the tag is rendered?

TL;DR yes, that's how you would do it and is how we customize for tags like `@param` and `@return`.

**Comment by thomthom** [comment]
Thanks. Thanks helps.

Follow up question, when do you use directives over tags? Seems like directives is a form of tags?

**Comment by lsegal** [comment]
Tags come in two forms:

1. Metadata -- use these when you want to add _data_ (just data) to an object.
2. Directives -- use these when you want to add _behavior_ to an object. Specifically, directives themselves are not stored on an object, they are _executed_ on an object. The output of a directive execution can be the addition of metadata, or it can be manipulation of parsing state (as in the case of `@!scope`), etc

**Comment by lsegal** [comment]
I am going to mark this as answered, feel free to re-open or create a new issue if there is more to ask!
